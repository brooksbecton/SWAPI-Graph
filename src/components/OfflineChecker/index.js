// @flow

/**
  Component monitors/handles network status and displays a toast when it detects an offline status. 
 */

import React, { Component } from "react";
import { message } from "antd";

type Props = {
  handleNetworkChange: ?(isOnline: boolean) => null
};

type State = {
  online: boolean
};

class OfflineChecker extends Component<Props, State> {
  constructor() {
    super();
    this.state = {
      online: true
    };
  }

  componentDidMount() {
    this.addOfflineListener();
  }

  componentWillUnmount() {
    this.removeNetworkListeners();
  }

  addOfflineListener = (): void => {
    window.addEventListener("offline", this.updateOnlineStatus);
  };
  removeNetworkListeners = (): void => {
    window.removeEventListener("offline", this.updateOnlineStatus);
  };

  updateOnlineStatus = (): void => {
    const isOnline: boolean = navigator.onLine;

    if (!isOnline) {
      if (this.props.handleNetworkChange) {
        this.props.handleNetworkChange(isOnline);
      }

      window.addEventListener("online", this.updateOnlineStatus);
    } else {
      window.removeEventListener("online", () => {});
    }

    this.setState({ online: isOnline });
  };

  render() {
    return (
      <div>{!this.state.online && message.warning("You are offline.")}</div>
    );
  }
}

export default OfflineChecker;
