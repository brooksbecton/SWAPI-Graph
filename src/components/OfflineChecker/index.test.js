import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import OfflineChecker from "./index";

Enzyme.configure({ adapter: new Adapter() });

describe("<OfflineChecker/>", () => {
  it("renders correctly if online", () => {
    expect(shallow(<OfflineChecker />)).toMatchSnapshot();
  });
});
