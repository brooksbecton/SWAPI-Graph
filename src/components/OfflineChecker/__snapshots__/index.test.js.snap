// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfflineChecker/> renders correctly if offline 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OfflineChecker />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": false,
    },
    "ref": null,
    "rendered": false,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": false,
      },
      "ref": null,
      "rendered": false,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<OfflineChecker/> renders correctly if online 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OfflineChecker />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": false,
    },
    "ref": null,
    "rendered": false,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": false,
      },
      "ref": null,
      "rendered": false,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
